File: packages/adl-agent/implementations/anthropic-py/agent.py
"""
ADL Explainer Agent -- Anthropic Python SDK implementation.

Uses the anthropic package for the Claude API tool-use loop.
"""

from __future__ import annotations

import json
from typing import Any

import anthropic

from tools import execute_tool

SYSTEM_PROMPT = (
    "You are the ADL Spec Explainer, an expert on the Agent Definition Language (ADL).\n"
    "\n"
    "You help users:\n"
    "- Understand ADL concepts (the passport model, tools, permissions, data classification, "
    "profiles, lifecycle, security, extensions)\n"
    "- Validate their ADL documents and explain any errors in plain language\n"
    "- Explore example ADL documents\n"
    "- Navigate the ADL specification\n"
    "- Compare ADL with other agent description formats (A2A, MCP, OpenAPI, Agent Protocol)\n"
    "\n"
    "Always be clear, accurate, and helpful. When explaining validation errors, provide "
    "actionable guidance on how to fix them. Use the available tools to look up information "
    "rather than relying on memory alone.\n"
    "\n"
    "ADL documents are YAML or JSON files that describe AI agents -- like a passport that "
    "travels with the agent, declaring its identity, capabilities, permissions, and trust signals."
)

TOOLS: list[dict[str, Any]] = [
    {
        "name": "explain_concept",
        "description": (
            "Explain an ADL concept such as the passport model, tools, permissions, "
            "data classification, profiles, lifecycle, security, or extensions."
        ),
        "input_schema": {
            "type": "object",
            "properties": {
                "concept": {
                    "type": "string",
                    "description": "The ADL concept to explain",
                },
            },
            "required": ["concept"],
        },
    },
    {
        "name": "validate_document",
        "description": (
            "Validate an ADL document provided as YAML or JSON and return structured "
            "results with errors explained in plain language."
        ),
        "input_schema": {
            "type": "object",
            "properties": {
                "document": {
                    "type": "string",
                    "description": "The ADL document content as YAML or JSON string",
                },
            },
            "required": ["document"],
        },
    },
    {
        "name": "show_example",
        "description": "Return an example ADL document by category: minimal, production, or with-tools.",
        "input_schema": {
            "type": "object",
            "properties": {
                "category": {
                    "type": "string",
                    "enum": ["minimal", "production", "with-tools"],
                    "description": "The example category to retrieve",
                },
            },
            "required": ["category"],
        },
    },
    {
        "name": "get_spec_section",
        "description": "Retrieve information about a specific section of the ADL specification.",
        "input_schema": {
            "type": "object",
            "properties": {
                "section": {
                    "type": "string",
                    "description": "The spec section or topic to look up",
                },
            },
            "required": ["section"],
        },
    },
    {
        "name": "compare_formats",
        "description": (
            "Compare ADL with another agent description format such as A2A, MCP, "
            "OpenAPI, or Agent Protocol."
        ),
        "input_schema": {
            "type": "object",
            "properties": {
                "format": {
                    "type": "string",
                    "enum": ["a2a", "mcp", "openai", "agent_protocol"],
                    "description": "The format to compare against ADL",
                },
            },
            "required": ["format"],
        },
    },
]


def run_agent_turn(
    client: anthropic.Anthropic,
    messages: list[dict[str, Any]],
    model: str = "claude-sonnet-4-20250514",
    max_tokens: int = 4096,
) -> str:
    """Run a single agent turn: send messages, handle tool calls, return final text."""
    response = client.messages.create(
        model=model,
        max_tokens=max_tokens,
        system=SYSTEM_PROMPT,
        tools=TOOLS,
        messages=messages,
    )

    while response.stop_reason == "tool_use":
        tool_use_blocks = [b for b in response.content if b.type == "tool_use"]

        messages.append({"role": "assistant", "content": response.content})

        tool_results = []
        for block in tool_use_blocks:
            result = execute_tool(block.name, block.input)
            tool_results.append({
                "type": "tool_result",
                "tool_use_id": block.id,
                "content": result,
            })

        messages.append({"role": "user", "content": tool_results})

        response = client.messages.create(
            model=model,
            max_tokens=max_tokens,
            system=SYSTEM_PROMPT,
            tools=TOOLS,
            messages=messages,
        )

    messages.append({"role": "assistant", "content": response.content})

    text_parts = [b.text for b in response.content if b.type == "text"]
    return "\n".join(text_parts)

File: packages/adl-agent/src/agent.ts
/**
 * Agent core — Anthropic SDK agent loop.
 * Re-exports shared tools/system prompt and provides the Anthropic-specific agent turn.
 */

import Anthropic from "@anthropic-ai/sdk";
import type {
  MessageParam,
  ToolResultBlockParam,
} from "@anthropic-ai/sdk/resources/messages";
import { SYSTEM_PROMPT, TOOLS, executeTool, type ToolDefinition } from "./shared.js";

// Re-export for backward compatibility
export { SYSTEM_PROMPT, TOOLS as TOOL_DEFINITIONS, executeTool };

/**
 * Map shared ToolDefinitions to Anthropic API Tool format.
 */
function toAnthropicTools(tools: ToolDefinition[]): Anthropic.Tool[] {
  return tools.map((t) => ({
    name: t.name,
    description: t.description,
    input_schema: {
      type: "object" as const,
      properties: t.parameters.properties,
      required: t.parameters.required,
    },
  }));
}

const ANTHROPIC_TOOLS = toAnthropicTools(TOOLS);

export interface AgentConfig {
  model?: string;
  maxTokens?: number;
}

/**
 * Run a single turn of the agent — send a message, handle tool calls, return the final text.
 */
export async function runAgentTurn(
  client: Anthropic,
  messages: MessageParam[],
  config?: AgentConfig,
): Promise<string> {
  const model = config?.model ?? "claude-sonnet-4-20250514";
  const maxTokens = config?.maxTokens ?? 4096;

  let response = await client.messages.create({
    model,
    max_tokens: maxTokens,
    system: SYSTEM_PROMPT,
    tools: ANTHROPIC_TOOLS,
    messages,
  });

  while (response.stop_reason === "tool_use") {
    const toolUseBlocks = response.content.filter(
      (block): block is Anthropic.Messages.ToolUseBlock =>
        block.type === "tool_use",
    );

    messages.push({ role: "assistant", content: response.content });

    const toolResults: ToolResultBlockParam[] = toolUseBlocks.map((block) => ({
      type: "tool_result" as const,
      tool_use_id: block.id,
      content: executeTool(block.name, block.input as Record<string, unknown>),
    }));

    messages.push({ role: "user", content: toolResults });

    response = await client.messages.create({
      model,
      max_tokens: maxTokens,
      system: SYSTEM_PROMPT,
      tools: ANTHROPIC_TOOLS,
      messages,
    });
  }

  messages.push({ role: "assistant", content: response.content });

  const textBlocks = response.content.filter(
    (block): block is Anthropic.Messages.TextBlock => block.type === "text",
  );

  return textBlocks.map((block) => block.text).join("\n");
}
